File: implementations/react/src/patterns/behavioral/state/context.ts
import { NewState } from "./new-state";
import type { TicketState } from "./state";

export class TicketContext {
  state: TicketState;

  title = "";
  tasks: Array<{ id: string; title: string; done: boolean }> = [];
  closedAt: Date | null = null;

  draftTitle = "";
  draftTasks: string[] = [];


  constructor() {
    this.state = new NewState(this);
  }

  setState(state: TicketState) {
    this.state = state;
  }

  request() {
    this.state.handleRequest();
  }


  
  commitDraft() {
    const t = this.draftTitle.trim();
    const list = this.draftTasks.map(s => s.trim()).filter(Boolean);
    if (!t) throw new Error("Title is required");
    if (list.length === 0) throw new Error("Add at least one task");
    this.title = t;
    this.tasks = list.map((txt, i) => ({ id: String(i + 1), title: txt, done: false }));
    this.closedAt = null;
  }
  checkNextTask() {
    const next = this.tasks.find(x => !x.done);
    if (!next) return null;
    next.done = true;
    return next;
  }
  allDone() {
    return this.tasks.length > 0 && this.tasks.every(x => x.done);
  }
  reset() {
    this.title = "";
    this.tasks = [];
    this.closedAt = null;
    this.draftTitle = "";
    this.draftTasks = [];
    this.setState(new NewState(this));
  }
}

File: implementations/flutter/design_patterns_flutter/lib/patterns/behavioral/state/context.dart
import 'package:design_patterns_flutter/patterns/behavioral/state/models/task.dart';
import 'package:design_patterns_flutter/patterns/behavioral/state/new_state.dart';
import 'package:design_patterns_flutter/patterns/behavioral/state/state.dart';

class TicketContext {
  late TicketState state;

  String title = '';
  List<Task> tasks = [];
  DateTime? closedAt;

  String draftTitle = '';
  List<String> draftTasks = [];

  TicketContext() {
    state = NewState(this);
  }

  void setState(TicketState s) {
    state = s;
  }

  void request() {
    state.handleRequest();
  }

  void commitDraft() {
    final t = draftTitle.trim();
    final list = draftTasks
        .map((s) => s.trim())
        .where((s) => s.isNotEmpty)
        .toList();
    if (t.isEmpty) throw Exception('Title is required');
    if (list.isEmpty) throw Exception('Add at least one task');
    title = t;
    tasks = [
      for (int i = 0; i < list.length; i++)
        Task(id: '${i + 1}', title: list[i]),
    ];
    closedAt = null;
  }

  Task? checkNextTask() {
    final next = tasks.firstWhere(
      (x) => !x.done,
      orElse: () => Task(id: '', title: '', done: true),
    );
    if (next.id.isEmpty) return null;
    next.done = true;
    return next;
  }

  bool allDone() => tasks.isNotEmpty && tasks.every((x) => x.done);

  void reset() {
    title = '';
    tasks = [];
    closedAt = null;
    draftTitle = '';
    draftTasks = [];
    setState(NewState(this));
  }
}
